File: src/main/scala/com/lmmmowi/leetcode/scala/p1/Solution.scala
package com.lmmmowi.leetcode.scala.p1

import scala.collection.mutable

/**
  * @Author: mowi
  * @Date: 2019-05-11
  * @Description: 1.两数之和[https://leetcode-cn.com/problems/two-sum/]
  */
object Solution {

    def twoSum(nums: Array[Int], target: Int): Array[Int] = {
        val map = new mutable.HashMap[Int, Int]()

        for (i <- nums.indices) {
            val foundIndex = map.get(target - nums(i))
            if (foundIndex.isDefined) {
                return Array(i, foundIndex.get)
            }
            else {
                map.put(nums(i), i)
            }
        }

        null
    }
}

File: src/main/java/com/lmmmowi/leetcode/p1/Solution.java
package com.lmmmowi.leetcode.p1;

import java.util.HashMap;
import java.util.Map;

/**
 * @Author: mowi
 * @Date: 2019-05-11
 * @Description: 1.两数之和[https://leetcode-cn.com/problems/two-sum/]
 */
public class Solution {

    public int[] twoSum(int[] nums, int target) {
        int size = (nums.length / 3 + 1) * 4;
        Map<Integer, Integer> map = new HashMap<Integer, Integer>(size);

        for (int i = 0; i < nums.length; i++) {
            Integer index = map.get(target - nums[i]);
            if (index != null) {
                return new int[]{i, index};
            } else {
                map.put(nums[i], i);
            }
        }

        return null;
    }
}
